### PR TITLE
[#596] Add checks for service and deployments creation and removing

### LINKED ISSUE #596: Add waits in undeploy cycle
Now we are not waiting during service and deployment removal. We have to add wait to be sure that model deployment will be removed only after model service.

Now this problem generates error:
```
Exception: Got error from server: 'Cannot find deployment for model service <bound method ModelService.__str__ of ModelService for model edi-test-model version 1.2>'
```

### PR DESCRIPTION
**Pull request by kirillmakhonin**
This closes #596

### CODE REVIEW

**Comment by kirillmakhonin** [comment]
bot, test all

**Comment by kirillmakhonin** [comment]
bot, test all

**Comment by kirillmakhonin** [comment]
bot, test all

**Comment by kirillmakhonin** [comment]
bot, test all

**Comment by kirillmakhonin** [comment]
bot, test app

**Comment by kirillmakhonin** [comment]
bot, test all

**Comment by kirillmakhonin** [comment]
bot, test app

**Comment by kirillmakhonin** [comment]
bot, test all

**Comment by kirillmakhonin** [comment]
I have to remove resources on fail, add custom exceptions

File: legion/legion/k8s/enclave.py [hunk 1]
```diff
@@ -388,6 +388,16 @@ def deploy_model(self, image, model_iam_role=None, count=1, livenesstimeout=2, r
             body=deployment,
             namespace=self.namespace)
 
+        deployment_ready = legion.utils.ensure_function_succeed(
+            lambda: image_meta_information.k8s_name in [
+                item.metadata.name
+                for item in extensions_v1beta1.list_namespaced_deployment(self.namespace).items],
+            10, 1, boolean_check=True
```
Comment (dsuslov) [inline legion/legion/k8s/enclave.py]: @kirillmakhonin, move magic numbers to configuration/profiles

File: legion/legion/k8s/enclave.py [hunk 2]
```diff
@@ -409,6 +419,16 @@ def deploy_model(self, image, model_iam_role=None, count=1, livenesstimeout=2, r
             body=service,
             namespace=self.namespace)
 
+        service_ready = legion.utils.ensure_function_succeed(
+            lambda: image_meta_information.k8s_name in [
+                item.metadata.name
+                for item in core_v1api.list_namespaced_service(self.namespace).items],
+            10, 1, boolean_check=True
```
Comment (dsuslov) [inline legion/legion/k8s/enclave.py]: @kirillmakhonin, move magic numbers to configuration/profiles

File: legion/legion/k8s/enclave.py [hunk 3]
```diff
@@ -388,6 +388,16 @@ def deploy_model(self, image, model_iam_role=None, count=1, livenesstimeout=2, r
...
 
+        deployment_ready = legion.utils.ensure_function_succeed(
+            lambda: image_meta_information.k8s_name in [
+                item.metadata.name
+                for item in extensions_v1beta1.list_namespaced_deployment(self.namespace).items],
+            10, 1, boolean_check=True
+        )
+
+        if not deployment_ready:
+            raise Exception('Cannot get deployment info')
```
Comment (dsuslov) [inline legion/legion/k8s/enclave.py]: @kirillmakhonin use custom exception type to make custom handling logic possible in future

File: legion/legion/k8s/enclave.py [hunk 4]
```diff
@@ -409,6 +419,16 @@ def deploy_model(self, image, model_iam_role=None, count=1, livenesstimeout=2, r
...
 
+        service_ready = legion.utils.ensure_function_succeed(
+            lambda: image_meta_information.k8s_name in [
+                item.metadata.name
+                for item in core_v1api.list_namespaced_service(self.namespace).items],
+            10, 1, boolean_check=True
+        )
+
+        if not service_ready:
+            raise Exception('Cannot get deployment info')
```
Comment (dsuslov) [inline legion/legion/k8s/enclave.py]: @kirillmakhonin use custom exception type to make custom handling logic possible in future

File: legion/legion/k8s/services.py [hunk 5]
```diff
@@ -371,13 +371,33 @@ def delete(self, grace_period_seconds=0):
...
 
+        service_deleted = ensure_function_succeed(
+            lambda: self.k8s_service.metadata.name not in [
+                item.metadata.name
+                for item in core_v1api.list_namespaced_service(self.namespace).items],
+            10, 1, boolean_check=True
+        )
+
+        if not service_deleted:
+            raise Exception('Cannot delete service')
```
Comment (dsuslov) [inline legion/legion/k8s/services.py]: @kirillmakhonin use custom exception type to make custom handling logic possible in future

File: legion/legion/k8s/services.py [hunk 6]
```diff
@@ -371,13 +371,33 @@ def delete(self, grace_period_seconds=0):
...
 
+        deployment_deleted = ensure_function_succeed(
+            lambda: self.deployment.metadata.name not in [
+                item.metadata.name
+                for item in api_instance.list_namespaced_deployment(self.namespace).items],
+            10, 1, boolean_check=True
+        )
+
+        if not deployment_deleted:
+            raise Exception('Cannot delete deployment')
```
Comment (dsuslov) [inline legion/legion/k8s/services.py]: @kirillmakhonin use custom exception type to make custom handling logic possible in future

File: legion/legion/k8s/services.py [hunk 7]
```diff
@@ -371,13 +371,33 @@ def delete(self, grace_period_seconds=0):
         core_v1api.delete_namespaced_service(name=self.k8s_service.metadata.name, body=body,
                                              namespace=self.namespace)
 
+        service_deleted = ensure_function_succeed(
+            lambda: self.k8s_service.metadata.name not in [
+                item.metadata.name
+                for item in core_v1api.list_namespaced_service(self.namespace).items],
+            10, 1, boolean_check=True
```
Comment (dsuslov) [inline legion/legion/k8s/services.py]: @kirillmakhonin, move magic numbers to configuration/profiles

File: legion/legion/k8s/services.py [hunk 8]
```diff
@@ -371,13 +371,33 @@ def delete(self, grace_period_seconds=0):
...
         api_instance.delete_namespaced_deployment(name=self.deployment.metadata.name,
                                                   namespace=self.namespace, body=body,
                                                   grace_period_seconds=grace_period_seconds,
                                                   propagation_policy='Background')
 
+        deployment_deleted = ensure_function_succeed(
+            lambda: self.deployment.metadata.name not in [
+                item.metadata.name
+                for item in api_instance.list_namespaced_deployment(self.namespace).items],
+            10, 1, boolean_check=True
```
Comment (dsuslov) [inline legion/legion/k8s/services.py]: @kirillmakhonin, move magic numbers to configuration/profiles

**Comment by dsuslov** [comment]
@kirillmakhonin, add exception handling in edi server where enclave.py methods might throw exceptions

**Comment by kirillmakhonin** [comment]
bot, test all

**Comment by kirillmakhonin** [comment]
bot, test all

**Comment by kirillmakhonin** [comment]
bot, test all

**Comment by kirillmakhonin** [comment]
bot, test app

**Comment by kirillmakhonin** [comment]
bot, test app

**Comment by kirillmakhonin** [comment]
bot, test app

**Comment by kirillmakhonin** [comment]
bot, test app

File: deploy/ansible/roles/legion_enclave/defaults/main.yml [hunk 9]
```diff
@@ -6,4 +6,6 @@ domain_delimiter: "-"
 namespace: "{{ enclave }}"
 source_secret_name: "{{ base_domain }}-tls"
 airflow_db_schema: "{{ enclave }}"
-airflow_db_enclave_user: "{{ enclave }}"
\ No newline at end of file
+airflow_db_enclave_user: "{{ enclave }}"
+k8s_api_retries_count: 10
```
Comment (dsuslov) [inline deploy/ansible/roles/legion_enclave/defaults/main.yml]: @kirillmakhonin please give right and clear names for configuration items.

k8s_api_retries_count -> k8s_api_retry_number_max_limit
k8s_api_retries_wait -> k8s_api_retry_delay_sec

File: legion/legion/k8s/services.py [hunk 10]
```diff
@@ -371,13 +404,36 @@ def delete(self, grace_period_seconds=0):
...
 
+        retries = int(os.getenv(*legion.config.K8S_API_RETRIES_COUNT))
+        retry_timeout = int(os.getenv(*legion.config.K8S_API_RETRIES_WAIT))
+
+        service_deleted = ensure_function_succeed(
+            lambda: self.check_service_is_deleted(),
+            retries, retry_timeout, boolean_check=True
+        )
+
+        if not service_deleted:
```
Comment (dsuslov) [inline legion/legion/k8s/services.py]: @kirillmakhonin how do you distinguish 2 different cases 1) service just was not deleted 2)  the call self.check_service_is_deleted() was not completed by some reason related to k8s ?

File: legion/legion/k8s/services.py [hunk 11]
```diff
@@ -351,6 +352,38 @@ def scale(self, new_scale):
...
+
+        :return: bool -- has deployment been deleted
+        """
+        client = legion.k8s.utils.build_client()
+        extension_api = kubernetes.client.ExtensionsV1beta1Api(client)
+
+        actual_deployments = [
+            item.metadata.name
+            for item
+            in extension_api.list_namespaced_deployment(self.namespace).items
```
Comment (dsuslov) [inline legion/legion/k8s/services.py]: @kirillmakhonin add retries for extension_api.list_namespaced_deployment(self.namespace)

File: legion/legion/k8s/utils.py [hunk 12]
```diff
@@ -207,8 +207,14 @@ def get_meta_from_docker_labels(labels):
     :type labels: dict[str, Any]
     :return: tuple[str, dict[str, str], str, str] -- k8s_name name, labels in DNS-1123 format, model id and version
     """
-    model_id = labels[legion.containers.headers.DOMAIN_MODEL_ID]
-    model_version = labels[legion.containers.headers.DOMAIN_MODEL_VERSION]
+    model_id = labels.get(legion.containers.headers.DOMAIN_MODEL_ID)
+    model_version = labels.get(legion.containers.headers.DOMAIN_MODEL_VERSION)
+
+    if not model_id or not model_version:
+        raise Exception('Legion docker labels for model image are missed: {}'.format(
```
Comment (dsuslov) [inline legion/legion/k8s/utils.py]: @kirillmakhonin too general type of exception

**Comment by kirillmakhonin** [comment]
bot, test app

**Comment by kirillmakhonin** [comment]
bot, test all

**Pull request merged**